lib/ts/recipe/session/sessionAuth.tsx

```diff
@@ -47,125 +49,96 @@ const SessionAuth: React.FC<Props> = ({ children, ...props }) => {
 
     const session = useRef(Session.getInstanceOrThrow());
 
-    const setInitialContext = useCallback(async (): Promise<void> => {
-        const sessionExists = await doesSessionExist();
+    // on mount
+    useEffect(() => {
+        let cancelUseEffect = false;
+        async function setInitialContextAndMaybeRedirect() {
+            // setting context from parent or manually
+            let toSetContext: SessionContextType;
+
+            if (hasParentProvider(parentSessionContext)) {
+                toSetContext = parentSessionContext;
+            } else {
+                const sessionExists = await doesSessionExist();
+
+                if (sessionExists === false) {
+                    toSetContext = {
+                        doesSessionExist: sessionExists,
+                        jwtPayload: {},
+                        userId: "",
+                    };
+                } else {
+                    const jwtPayload = await getJWTPayloadSecurely();
+                    const userId = await getUserId();
+                    toSetContext = {
+                        doesSessionExist: sessionExists,
+                        jwtPayload,
+                        userId,
+                    };
+                }
+            }
 
-        if (sessionExists === false) {
-            return setContext({
-                doesSessionExist: sessionExists,
-                jwtPayload: {},
-                userId: "",
-            });
-        }
+            // if this component is unmounting, or the context has already
+            // been set, then we don't need to proceed...
+            if (cancelUseEffect || context !== undefined) {
+                return;
+            }
 
-        const [jwtPayload, userId] = await Promise.all([getJWTPayloadSecurely(), getUserId()]);
+            if (!toSetContext.doesSessionExist && props.requireAuth === true) {
+                props.redirectToLogin();
+            } else {
+                setContext(toSetContext);
+            }
+        }
 
-        return setContext({
-            doesSessionExist: sessionExists,
-            jwtPayload,
-            userId,
-        });
+        setInitialContextAndMaybeRedirect();
+        return () => {
+            cancelUseEffect = true;
+        };
     }, []);
```

Comment:
What if the `requireAuth` prop changes? We are not going to execute this logic, despite that it changes the behaviour (user won't get redirected when this prop changes).

Reply:
That's reasonable. I'll add this.